### PR TITLE
[JUJU-1479] Check lxd version

### PR DESCRIPTION
**Pull request by ycliuhw**
This PR checks if LXD version >= 5.2(if the model's cloud is LXD) when we upgrade a controller/model to 3.0.0+ or migrate a model to a 3.0.0+ controller.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [ ] ~Comments saying why design decisions were made~
- [x] Go unit tests, with comments saying what you're testing
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

```console
# upgrade

$ lxd --version
5.0.0

$ juju models -c k1 --format json | jq '.models[] | {name,"agent-version"}'
{
  "name": "admin/controller",
  "agent-version": "2.9.33.1"
}
{
  "name": "admin/default",
  "agent-version": "2.9.33.1"
}

$ juju version --all
version: 3.0.0-ubuntu-amd64
git-commit: d159efac923ce6d424c021c105b1f7e130ffcdd2
git-tree-state: dirty
compiler: gc

$ juju upgrade-controller --build-agent
no prepackaged agent binaries available, using local agent binary 3.0.0.2 (built from source)
ERROR cannot upgrade to "3.0.0.2" due to issues with these models:
"admin/controller":
- LXD version has to be "5.2.0" at least, but current version is "5.0.0"
"admin/default":
- LXD version has to be "5.2.0" at least, but current version is "5.0.0"

# migrate

$ juju add-model m1 -c k1
Added 'm1' model on localhost/localhost with credential 'localhost' for user 'admin'

$ juju models -c k1 --format json | jq '.models[] | {name,"agent-version"}'
{
  "name": "admin/controller",
  "agent-version": "2.9.33.2"
}
{
  "name": "admin/default",
  "agent-version": "2.9.33.1"
}
{
  "name": "admin/m1",
  "agent-version": "2.9.33.2"
}

$ juju models -c target --format json | jq '.models[] | {name,"agent-version"}'
{
  "name": "admin/controller",
  "agent-version": "3.0.0.1"
}
{
  "name": "admin/default",
  "agent-version": "3.0.0.1"
}

$ juju migrate k1:m1 target
ERROR source prechecks failed: cannot migrate to controller ("3.0.0.1") due to issues:
"admin/m1":
- LXD version has to be "5.2.0" at least, but current version is "5.0.0"

```

## Documentation changes

No

## Bug reference

https://bugs.launchpad.net/juju/+bug/1981955

### CODE REVIEW

**Review by wallyworld — COMMENTED** [review]
I'm not sure the lxd check is totally correct. It handles the case where the cloud itself is lxd. But also, any cloud can create a lxd container on a machine host. We should probably look at  all the machines in the model as well (like we do for checking series) and see if they have lxd installed, and if it is installed, is it the right version. We could do this extra check as a follow up PR so as not to block this one.

Also, I can see migration and upgrade checks - but we also want to check at bootstrap time as well.

**Comment by wallyworld** [comment]
> I'm not sure the lxd check is totally correct. It handles the case where the cloud itself is lxd. But also, any cloud can create a lxd container on a machine host. We should probably look at all the machines in the model as well (like we do for checking series) and see if they have lxd installed, and if it is installed, is it the right version. We could do this extra check as a follow up PR so as not to block this one.
> 
> Also, I can see migration and upgrade checks - but we also want to check at bootstrap time as well.

Ah, this is for 2.9 branch. The bootstrap check will be for 3.0

**Comment by ycliuhw** [comment]
> > I'm not sure the lxd check is totally correct. It handles the case where the cloud itself is lxd. But also, any cloud can create a lxd container on a machine host. We should probably look at all the machines in the model as well (like we do for checking series) and see if they have lxd installed, and if it is installed, is it the right version. We could do this extra check as a follow up PR so as not to block this one.
> > Also, I can see migration and upgrade checks - but we also want to check at bootstrap time as well.
> 
> Ah, this is for 2.9 branch. The bootstrap check will be for 3.0

Yeah, `JUJU-1392` is for that. Current PR is only for upgrade and migrate.

File: provider/lxd/server.go [hunk 1]
```diff
@@ -374,22 +375,32 @@ func (s *serverFactory) Clock() clock.Clock {
...
+	versionParts := strings.Split(s, ".")
 	if len(versionParts) < 2 {
-		return fmt.Sprintf("LXD API version %q: expected format <major>.<minor>", version), false
+		return version.Zero, errors.NewNotValid(nil, fmt.Sprintf("LXD API version %q: expected format <major>.<minor>", s))
 	}
-
 	major, err := strconv.Atoi(versionParts[0])
 	if err != nil {
-		return fmt.Sprintf("LXD API version %q: unexpected major number: %v", version, err), false
+		return version.Zero, errors.NotValidf("major %v", versionParts[0])
```
Comment (wallyworld) [inline provider/lxd/server.go:386]: "major version number %q"

File: provider/lxd/server.go [hunk 2]
```diff
@@ -374,22 +375,32 @@ func (s *serverFactory) Clock() clock.Clock {
...
 	}
+	minor, err := strconv.Atoi(versionParts[1])
+	if err != nil {
+		return version.Zero, errors.NotValidf("minor %v", versionParts[1])
```
Comment (wallyworld) [inline provider/lxd/server.go:390]: "minor version number %q"

File: upgrades/upgradevalidation/validation.go [hunk 3]
```diff
@@ -279,3 +284,39 @@ func checkMongoVersionForControllerModel(modelUUID string, pool StatePool, st St
...
+		}
+		serverVersion := server.ServerVersion()
+		currentVer, err := lxd.ParseAPIVersion(serverVersion)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		logger.Debugf("current LXD version %q, min LXD version %q", currentVer, minVer)
+		if currentVer.Compare(minVer) < 0 {
+			return NewBlocker(
+				"LXD version has to be %q at least, but current version is %q",
```
Comment (wallyworld) [inline upgrades/upgradevalidation/validation.go:316]: `LXD version has to be at least %q, but current version is only %q`

**Review by wallyworld — APPROVED** [review]
Thank you, just a couple of minor error message tweaks

**Comment by ycliuhw** [comment]
> I'm not sure the lxd check is totally correct. It handles the case where the cloud itself is lxd. But also, any cloud can create a lxd container on a machine host. We should probably look at all the machines in the model as well (like we do for checking series) and see if they have lxd installed, and if it is installed, is it the right version. We could do this extra check as a follow up PR so as not to block this one.
> 
> Also, I can see migration and upgrade checks - but we also want to check at bootstrap time as well.

As discussed, we are not going to check the LXD installed on host machines. It's not worth spending too much time on this edge case.

**Pull request merged**